File: 2016/src/day22.go
package src

import (
    "fmt"
    "strings"
    "strconv"
    . "../utils"
)

type node struct {
    Pos   Point
    Size  int
    Used  int
    Avail int
}

type state22 struct {
    Empty Point
    Goal Point
}

func Solve22(input string) string {
    var ans1 int
    var ans2 int

    grid := map[Point]node{}
    nodes := []node{}
    datax := 0
    var empty node
    for l, line := range strings.Split(input, "\n") {
        if l < 2 {
            continue
        }

        splits := strings.FieldsFunc(line, func (r rune) bool {
            return r == ' ' || r == 'x' || r == 'y' || r == '-' || r == 'T'
        })

        x, _ := strconv.Atoi(splits[1])
        y, _ := strconv.Atoi(splits[2])
        size, _ := strconv.Atoi(splits[3])
        used, _ := strconv.Atoi(splits[4])
        avail, _ := strconv.Atoi(splits[5])

        n := node{Point{x, y}, size, used, avail}
        grid[Point{x, y}] = n

        if used == 0 {
            empty = n
        }

        if y == 0 && x > datax {
            datax = x
        }

        nodes = append(nodes, n)
    }

    for _, node := range nodes {
        if node.Used == 0 {
            continue
        }

        if node.Used <= empty.Avail {
            ans1++
        }
    }

    goal := Point{0, 0}
    visited := map[state22]bool{}
    toVisit := []state22{state22{empty.Pos, Point{datax, 0}}}

    bfs:
    for {
        newToVisit := []state22{}

        for _, state := range toVisit {
            if state.Goal == goal {
                break bfs
            }

            for _, offset := range []Point{Point{-1, 0}, Point{0, -1}, Point{1, 0}, Point{0, 1}} {
                newPos := state.Empty.Plus(offset)
                if node, exists := grid[newPos]; !exists || node.Size > 100 {
                    continue
                }

                var newState state22
                if newPos == state.Goal {
                    newState.Empty = state.Goal
                    newState.Goal = state.Empty
                } else {
                    newState.Empty = newPos
                    newState.Goal = state.Goal
                }

                if !visited[newState] {
                    visited[newState] = true
                    newToVisit = append(newToVisit, newState)
                }
            }
        }

        toVisit = newToVisit
        ans2++
    }

    return fmt.Sprintf("%d, %d", ans1, ans2)
}

File: 2017/src/day22.js
export default function solve22(input) {
    let ans1 = 0;
    let ans2 = 0;

    let lines = input.split("\n");
    let grid = new Set();

    let pos = [Math.floor(lines[0].length / 2), Math.floor(lines.length / 2)];

    for (let y = 0; y < lines.length; y++) {
        for (let x = 0; x < lines[y].length; x++) {
            if (lines[y][x] === "#") {
                grid.add([x, y].join(","));
            }
        }
    }

    function right(d) {
        return [-d[1], d[0]];
    }

    function left(d) {
        return [d[1], -d[0]];
    }

    let dir = [0, -1];

    let bursts = 10000;

    for (let i = 0; i < bursts; i++) {
        let posString = pos.join(",");
        if (grid.has(posString)) {
            dir = right(dir);
            grid.delete(posString);
        } else {
            dir = left(dir);
            grid.add(posString);
            ans1++;
        }

        pos = [pos[0] + dir[0], pos[1] + dir[1]];
    }

    grid = new Map();

    pos = [Math.floor(lines[0].length / 2), Math.floor(lines.length / 2)];
    dir = [0, -1];

    for (let y = 0; y < lines.length; y++) {
        for (let x = 0; x < lines[y].length; x++) {
            if (lines[y][x] === "#") {
                grid.set([x, y].join(","), "I");
            }
        }
    }

    bursts = 10000000;

    for (let i = 0; i < bursts; i++) {
        let posString = pos.join(",");

        let state = grid.get(posString);
        if (state === "W") {
            grid.set(posString, "I");
            ans2++;
        } else if (state === "I") {
            dir = right(dir);
            grid.set(posString, "F");
        } else if (state === "F") {
            dir = left(left(dir));
            grid.delete(posString);
        } else {
            dir = left(dir);
            grid.set(posString, "W");
        }

        pos = [pos[0] + dir[0], pos[1] + dir[1]];
    }

    return [ans1, ans2];
}
